py-polars/polars/internals/series/utils.py

```diff
@@ -0,0 +1,76 @@
+from __future__ import annotations
+
+import sys
+from typing import TYPE_CHECKING, Any, Callable
+
+import polars.internals as pli
+from polars.datatypes import DataType, dtype_to_ffiname
+
+if TYPE_CHECKING:
+    from polars.internals.type_aliases import Namespace
+    from polars.polars import PySeries
+
+    if sys.version_info >= (3, 10):
+        from typing import ParamSpec
+    else:
+        from typing_extensions import ParamSpec
+
+    P = ParamSpec("P")
+
+
+def expr(
+    namespace: Namespace | None = None,
+) -> Callable[[Callable[P, pli.Series]], Callable[P, pli.Series]]:
+    """
+    Dispatch Series method to an expression implementation.
+
+    Decorator.
+
+    Parameters
+    ----------
+    namespace : {'arr', 'cat', 'dt', 'str', 'struct'}
+        Expression namespace.
+
+    """
+
+    def decorator(func: Callable[P, pli.Series]) -> Callable[P, pli.Series]:
+        def wrapper(self: Any, *args: P.args, **kwargs: P.kwargs) -> pli.Series:
+            s = pli.wrap_s(self._s)
+            expr = pli.col(s.name)
+            if namespace is not None:
+                expr = getattr(expr, namespace)
+            f = getattr(expr, func.__name__)
+            return s.to_frame().select(f(*args, **kwargs)).to_series()
+
+        return wrapper  # type: ignore[return-value]
```

Comment:
very slick! what is `mypy` complaining about here? Is it because of `self: Any`?

Reply:
I was going to ask you for help! It's indeed the fact that I have to add the extra `self` parameter, since I am decorating a method. `mypy` says:

```
error: Incompatible return value type (got "Callable[[Any, **P], Series]", expected "Callable[P, Series]")  [return-value]
```

I've tried a bunch of things to no avail. Curious to hear if you have the fix here! I came to the conclusion that mypy simply doesn't support this construction 😬

Reply:
This seems like a really good use-case for `Concatenate`, but I'm also unable to satisfy `mypy`. I implemented the decorator as a class

```python
T = TypeVar(
    "T",
    ListNameSpace,
    StringNameSpace,
    StructNameSpace,
    DateTimeNameSpace,
    CatNameSpace,
    contravariant=True,
)
P = ParamSpec("P")


class call_expr:

    _namespace: Namespace

    def __init__(self, namespace: Namespace) -> None:
        self._namespace = namespace

    def __call__(
        self, method: Callable[Concatenate[T, P], pli.Series]
    ) -> Callable[Concatenate[T, P], pli.Series]:
        def wrapper(_self: T, *args: P.args, **kwargs: P.kwargs) -> pli.Series:
            s = pli.wrap_s(_self._s)
            expr = pli.col(s.name)
            if self._namespace is not None:
                expr = getattr(expr, self._namespace)
            f = getattr(expr, method.__name__)
            return s.to_frame().select(f(*args, **kwargs)).to_series()

        return wrapper
```

and I think this is *closer* to working, but I still get errors on `return wrapper`.

```
❯ mypy
polars/internals/series/utils.py:56: error: Incompatible return value type (got "Callable[[Arg(ListNameSpace, '_self'), **P], Series]", expected "Callable[[ListNameSpace, **P], Series]")  [return-value]
polars/internals/series/utils.py:56: note: This may be because "wrapper" has arguments named: "_self"
polars/internals/series/utils.py:56: error: Incompatible return value type (got "Callable[[Arg(StringNameSpace, '_self'), **P], Series]", expected "Callable[[StringNameSpace, **P], Series]")  [return-value]
polars/internals/series/utils.py:56: error: Incompatible return value type (got "Callable[[Arg(StructNameSpace, '_self'), **P], Series]", expected "Callable[[StructNameSpace, **P], Series]")  [return-value]
polars/internals/series/utils.py:56: error: Incompatible return value type (got "Callable[[Arg(DateTimeNameSpace, '_self'), **P], Series]", expected "Callable[[DateTimeNameSpace, **P], Series]")  [return-value]
polars/internals/series/utils.py:56: error: Incompatible return value type (got "Callable[[Arg(CatNameSpace, '_self'), **P], Series]", expected "Callable[[CatNameSpace, **P], Series]")  [return-value]
Found 5 errors in 1 file (checked 91 source files)
```

For some reason, `mypy` sees `Arg(ListNameSpace, '_self')` as a different thing than `ListNameSpace`

Reply:
My general thoughts are that `mypy` **should** recognize that 

```
Callable[[Arg(ListNameSpace, '_self'), **P], Series] == Callable[[ListNameSpace, **P], Series]
```

and eventually the `type: ignore` will become unnecessary (under my formulation, at least). But I will open an issue on the `mypy` side to confirm those two types are indeed equivalent.

Reply:
Interesting. I didn't consider `Concatenate` as we're not trying to add additional arguments to the method.

Overall, prefer the decorator factory approach that I have implemented. I think `ParamSpec` should support `P.self` in addition to `P.args` and `P.kwargs`. That would be the cleanest solution to this use case.

We'll just have to live with a `# type: ignore` for now 😃